### PR TITLE
make SIP not on by default

### PR DESCRIPTION
**Pull request by Maoni0**
I'm seeing with SIP the heap size can increase noticeably because we don't compact the SIP regions, for workload that rarely ever does blocking gen2s. so I'm making these not on by default but can be turned on with the GCEnableSpecialRegions config.

### CODE REVIEW

**Comment by Maoni0** [comment]
I also excluded a few methods that are only used for segments.

**Comment by Maoni0** [comment]
there are a few methods that are currently compiled for both segments and regions but are only used for segments so made those only compiled for segments.

File: src/coreclr/gc/gcpriv.h [hunk 1]
```diff
@@ -52,7 +52,7 @@ inline void FATAL_GC_ERROR()
 // This means any empty regions can be freely used for any generation. For
 // Server GC we will balance regions between heaps.
 // For now disable regions for StandAlone GC, NativeAOT and MacOS builds
-#if defined (HOST_64BIT) && !defined (BUILD_AS_STANDALONE) && !defined(__APPLE__) && !defined(FEATURE_NATIVEAOT)
+#if defined (HOST_64BIT) && defined (BUILD_AS_STANDALONE) && !defined(__APPLE__) && !defined(FEATURE_NATIVEAOT)
```
[the remark this answers is not in the document]
Comment (Maoni0) [reply src/coreclr/gc/gcpriv.h:55]: oops, thanks for catching that!!

File: src/coreclr/gc/gcpriv.h [hunk 2]
```diff
@@ -52,7 +52,7 @@ inline void FATAL_GC_ERROR()
 // This means any empty regions can be freely used for any generation. For
 // Server GC we will balance regions between heaps.
 // For now disable regions for StandAlone GC, NativeAOT and MacOS builds
-#if defined (HOST_64BIT) && !defined (BUILD_AS_STANDALONE) && !defined(__APPLE__) && !defined(FEATURE_NATIVEAOT)
```
Comment (cshung) [inline src/coreclr/gc/gcpriv.h:55]: This change is unintended?

**Review by cshung — APPROVED** [review]

File: src/coreclr/gc/gcpriv.h (hunk 2 above)
Comment (Maoni0) [reply src/coreclr/gc/gcpriv.h]: thanks. Manish also pointed out above, it's definitely not intended.

File: src/coreclr/gc/gcconfig.h [hunk 3]
```diff
@@ -105,6 +105,7 @@ class GCConfigStringHolder
     INT_CONFIG   (GCTotalPhysicalMemory,     "GCTotalPhysicalMemory",     NULL,                                0,                  "Specifies what the GC should consider to be total physical memory")                      \
     INT_CONFIG   (GCRegionRange,             "GCRegionRange",             NULL,                                0,                  "Specifies the range for the GC heap")                                                    \
     INT_CONFIG   (GCRegionSize,              "GCRegionSize",              NULL,                                4194304,            "Specifies the size for a basic GC region")                                               \
+    INT_CONFIG   (GCEnableSpecialRegions,    "GCEnableSpecialRegions",    NULL,                                0,                  "Specifies to enable special handling some regions like SIP")                             \
```
Comment (mangod9) [inline src/coreclr/gc/gcconfig.h:108]: do we want to document this, or is it only for certain workloads we know would benefit from SIP?

**Review by mangod9 — APPROVED** [review]

File: src/coreclr/gc/gcconfig.h (hunk 3 above)
Comment (Maoni0) [reply src/coreclr/gc/gcconfig.h:108]: for now it's just special workloads so I thought I'd leave it as an experimental thing (so not supported in runtimeconfig). we can decide whether to support in runtimeconmfig in the next release when we have experimented with it more. does that sound reasonable?

**Comment by Maoni0** [comment]
arm32 failure is due to linked known issue

**Pull request merged**